src/impression.js

```diff
@@ -47,15 +85,66 @@ export function maybeTrackImpression(win) {
     // avoid duplicate tracking.
     win.location.hash = '';
   }
+
   viewer.whenFirstVisible().then(() => {
-    invoke(win, clickUrl);
+    // TODO(@zhouyx) need test with a real response.
+    const promise = invoke(win, dev().assertString(clickUrl)).then(response => {
+      applyResponse(win, viewer, response);
+    });
+
+    // Timeout invoke promise after 8s and resolve trackImpressionPromise.
+    timerFor(win).timeoutPromise(TIMEOUT_VALUE, promise,
+        'timeout waiting for ad server response').catch(() => {
+        }).then(() => {
```

Comment:
This can be done with `resolveImpression(timerFor()...)`, which will skip the deferment from the `#then` block.